recipes-security/tboot/tboot-1.9.12/0010-Add-support-for-launching-64-bit-PE-kernels.patch

```diff
@@ -397,7 +398,7 @@ index 0000000..39b8422
 +     [lme] "i" (1 << _EFER_LME),
 +     [cr0_nopg] "i" (CR0_PE | CR0_MP | CR0_NE),
 +     [cr0] "i" (CR0_PE | CR0_MP | CR0_NE | CR0_PG),
-+     [cr4] "i" (CR4_DE | CR4_PAE | CR4_MCE | CR4_FXSR | CR4_XMM | CR4_SMXE),
```

Comment:
I'm actually not 100% convinced this section here is only for turning on PAE.  This is the last thing that happens before we jump back to a 64-bit PE.  In our case it's the xen.efi, but it really could be any PE.  Perhaps DeGraf wanted to leave the control register in a somewhat default (?) state (for an Intel vPRO chip) for the next PE to load after this call.

By the time we get to ```launch_kernel``` (and eventually this function) we've already done SENTER, which needs SMXE in CR4.  I guess since we aren't making any more SMX-required calls of the processor until we shutdown the platform (that I know of), and SMXE is guaranteed to be OR-d back into CR4 from a line in ```txt_shutdown```, ultimately it may not matter if we don't set it in CR4 before launching xen.efi the second time.  But I think this would represent a significant runtime difference in that on all versions prior to this, we have always run OpenXT with SMXE and these other vPRO flags in CR4, and taking this change would make that no longer the case.

Reply:
From memory, I think the code constructs page tables with the large bit set, so they are 2MB mappings.  CR4 needs PAE to support that.

tboot will boot without SMX/TXT - that is how I can get through here in QEMU.  I think you are supposed to check CPUID before you set it in CR4.  So this unilateral setting is wrong.  When SMX support is not detected, tboot shortcuts through most of its code.

My guess is de Graaf printed CR4 at some point and set it to that since it worked for the test system.  Since all vPro systems have the ~same capabilities, that was fine for real hardware.

Going through the CR4 bits:
I think these two made QEMU print out all the floating point registers (though maybe it was doing that regardless of these):
CR4_FXSR - Operating system support for FXSAVE and FXRSTOR instructions
CR4_XMM - Operating System Support for Unmasked SIMD Floating-Point Exceptions
tboot doesn't do floating point, so they don't need to be enabled, AFAIK.

CR4_DE - Debugging Extensions
These shouldn't be needed in normal operation, again AFAIK.
CR4_MCE Machine Check Exception
Not sure, but if this triggers interrupt generation, we don't want it set when transitioning to new code.

CR4_PAE - I think this is needed for [Physical Address Extension](https://en.wikipedia.org/wiki/Physical_Address_Extension) compatible page tables.

CR4_SMX - shouldn't be needed moving forward.

We're transitioning to a new PE, so it's a new software environment.  Xen has it's own code to handle CR4.

I see your point that this is a change in the processor CR4 configuration going forward.  I guess I could print CR4 with and without this change and see what changes in practice.  I guess my thought is that any new code needs to query CPUID and handle CR4 as it sees fit, so the new PE shouldn't be depending on the old value.

Real hardware that had SMX set will continue to have it set.  tboot is the only thing that does SMX/TXT instructions anyway, so having it flipped off outside of tboot shouldn't matter.  I expect Xen/Dom0 handle the other bits as necessary.

I'll report back in a bit with a finding on CR4 with and without this change.

Reply:
On an i7-8665U, when entering jump_pe_image, cr4 is 0x00044240

```
                    0x00044240
#define CR4_MCE     0x00000040 /* Machine check enable */
#define CR4_FXSR    0x00000200/* Fast FPU save/restore used by OS */
#define CR4_SMXE    0x00004000/* enable SMX */
#define CR4_OSXSAVE 0x00040000 /* enable xsave and xrestore */
```

So my change would drop CR4_DE and CR4_XMM (Called CR4_OSXMMEXCPT elsewhere).

CR4_DE
"Debugging Extensions"
"If set, enables debug register based breaks on I/O space access."

CR4_OSXMMEXCPT
"Operating System Support for Unmasked SIMD Floating-Point Exceptions"
"If set, enables unmasked SSE exceptions."

Xen __start_xen() sets CR4_OSXMMEXCPT during boot.

So CR4_DE is let unset.  From the SDM:
17.2.2 Debug Registers DR4 and DR5
Debug registers DR4 and DR5 are reserved when debug extensions are enabled (when the DE flag in control register CR4 is set) and attempts to reference the DR4 and DR5 registers cause invalid-opcode exceptions (#UD).  When debug extensions are not enabled (when the DE flag is clear), these registers are aliased to debug registers DR6 and DR7.

I think it's fine to leave this to Dom0/Xen to toggle as they need.  Xen looks to mainly reference CR4_DE for guest handling.  CR4_DE is included in:
```
/*
 * Bits which a PV guest can toggle in its view of cr4.  Some are loaded into
 * hardware, while some are fully emulated.
 */
#define PV_CR4_GUEST_MASK \
    (X86_CR4_TSD | X86_CR4_DE | X86_CR4_FSGSBASE | X86_CR4_OSXSAVE)
```

Linux itself clears CR4_DE for 64bit kernels (arch/x86/kernel/cpu/common.c).

Xen would also end up clearing out CR4_DE.  It starts with `mmu_cr4_features =  XEN_MINIMAL_CR4 = (X86_CR4_PGE | X86_CR4_PAE)` and  set_in_cr4() or-s in options.  It's in the later parts of __start_xen(), but this all means Xen doesn't rely on the inherited cr4.